rfc/20250718-enabled-support.md

```diff
@@ -0,0 +1,91 @@
+  # ...
+
+  lifecycle {
+    enabled = var.enable_server
+  }
+}
+```
+
+This is available not only to resources, but to modules too:
+
+```
+module "modcall" {
+  source = "./mod"
+
+  lifecycle {
+    enabled = var.env == "production"
+  }
+}
+```
+
+## Proposed Solution
+
+1. Raise errors if a resource is used while being disabled;
+1. Support for conditional enabling on:
+  1. Resources - Add a new field on the `lifecycle` block called `enabled`;
+  1. Modules - Add a new `lifecycle` block for Modules but only supporting the `enabled` field.
+
+
+## Technical Approach
+
+### `lifecycle` support on Modules
+
+Lifecycle block is currently not supported on `Modules`, but in our docs we're mentioning to users that it's reserved for future usage:
+
+> OpenTofu does not use the lifecycle argument. However, the lifecycle block is reserved for future versions.
+
+https://opentofu.org/docs/language/modules/syntax/#meta-arguments
+
+It would be the right timing to start to support it, but only for the `enabled` field.
+
+### Usage of `enabled` together with `for_each` and `count`.
+
+These three different arguments would behave similarly, but with different semantics.
+It could be argued that you want to enable a resource with 4 instances, but on our current code, you can already change that conditional by setting `count` to 0.
+
+I propose that we do not support them together. You can only use one of them, returning errors if we try to use the others together.
+
+Semantically, this feature would be used for single-instance resources or modules.
+
+### Errors when resource is not enabled
+
+If a resource is disabled, it shouldn't be possible to access the attributes since the resource
+is not on the graph. We prefer to loudly tell the user they're trying to access something that is not
+available. In order to have a good user experience while they want to enable/disable stuff, we
+offer a way to silently fail while trying to access that, by using the `try` function.
+
+### What happens when it's disabled
+
+Let's suppose we have a created a resource using the `lifecycle -> enabled` field and then we want to disable it.
+The behavior is going to be the same as if you wanted to destroy a resource:
+
+```
+# aws_instance.demo_vm_2 will be destroyed
+  - resource "aws_instance" "demo_vm_2" {
+      - ami                                  = "ami-07df274a488ca9195" -> null
+      - arn                                  = "arn:aws:ec2:eu-central-1:532199187081:instance/i-0b63433033e61d818" -> null
+      - associate_public_ip_address          = true -> null
+      - availability_zone                    = "eu-central-1b" -> null
+```
+
+### What types of variables are supported on the conditional
+
+There are a few types of values that cannot be used on this conditional:
+
+1. Unknown values, since they cannot be used until the apply phase
+1. Sensitive values
+1. Null values
+1. Conditionals that do not evaluate to true/bool values
```

Comment:
I would suggest including a "future considerations" section that talks about how we're adding `enabled` under `lifecycle` for now to avoid breaking uses of existing resource types or modules that already take an `enabled` argument with a different meaning.

We could then mention the possibility of improving this in a future language edition similar to what I described in https://github.com/opentofu/opentofu/issues/1306#issuecomment-2982113732.

My main goal for this is just so that someone who is wondering why we designed it this way can find the answer in this RFC along with a reminder of the idea we had to eventually improve it, which we could then link back to from a future RFC if we decide to improve this later.